### DISCUSSION TITLE
How does "CPU Usage" and "Memory Usage" work for a running job?

### DISCUSSION THREAD

**Discussion by iamumairayub** [jhuckaby/Cronicle#695] [Q&A] [answered]
So I have a Python script running inside the Shell Plugin and its running really well.

I just cant get my head around how Cronicle calculates those stats.

![screenshot taken from cronicle](https://github.com/jhuckaby/Cronicle/assets/13570295/88476324-bce5-4976-9950-10f163f5baf9)
![screenshot from MobaXterm](https://github.com/jhuckaby/Cronicle/assets/13570295/9eede598-36cf-4fbc-a51c-0c8c0de70076)

(both above screenshots were taken at exact same time)

Actually on server, overall CPU usage is 4% and overall RAM is 10GB/30GB used.

Regarding CPU, why its showing 13.99% even though total server CPU usage is 4%.
Regarding RAM, why the graph is all red? its showing 1.6GB so it should show graphics as per 1.6GB out of 30GB RAM.

**Comment by jhuckaby** [comment]
Cronicle calculates stats by executing `/bin/ps -eo "ppid pid %cpu rss"` every 10 seconds, and grepping ONLY your own running processes for your jobs (it also considers child processes if your jobs spawned any, then adding it all together). So it's only showing resources used by your job and your job's process and child processes, and not the entire server as a whole.

CPU is calculated as *percentage of one single CPU core* (this is also how utilities like `top` calculate it).  So if you have an 8-core CPU, then a job taking "100%" CPU is actually only ONE of those 8 cores.  I suspect your little task bar widget you are comparing to considers all cores maxed as being 100%, instead of one core.  Or maybe it could be showing you a 1-minute average.  I don't know anything about that widget.  But Cronicle is based on `/bin/ps` which should be accurate.

Your memory pie chart is red because your job is using over 1 GB RAM (again, this is based solely on your own job processes, not your server as a whole), and Cronicle is configured to use 1 GB as a "maximum" for the visual pie fill percent calculation.  You can change this by setting the [job_memory_max](https://github.com/jhuckaby/Cronicle/blob/master/docs/Configuration.md#job_memory_max) config prop in your `/opt/cronicle/conf/config.json` file.

Hope this helps.

**Reply by iamumairayub** [reply]
`job_memory_max` is already set to 0.
anyways thanks for nice explanation.

**Comment by jhuckaby** [comment, marked as answer]
> `job_memory_max` is already set to 0.

Yup, I understand, but Cronicle still needs some value to base the pie chart "maximum fill" level on, so if `job_memory_max` is undefined or zero, it defaults to 1 GB, just for the purpose of the UI pie.

**Reply by iamumairayub** [reply]
My server RAM is 30GB so if I set `job_memory_max` to 30GB will that pie chart show graphics as I want? means if particular job is taking 10GB of RAM it will show half of pie chart colored? (I cannot experiment this now as I have important long-running jobs already running)

How do I get to show same pie chart for CPU if I have 16 cores CPU?

**Reply by jhuckaby** [reply]
> My server RAM is 30GB so if I set job_memory_max to 30GB will that pie chart show graphics as I want? means if particular job is taking 10GB of RAM it will show half of pie chart colored?

Errr... yes, but it would show as 1/3 filled, not half.  10 GB / 30 GB is a third.

> How do I get to show same pie chart for CPU if I have 16 cores CPU?

Sorry, that one is hard-coded to show 100% of the pie filled as being equal to one CPU core.  Same as `top` and `ps`.

Cronicle v2 will show these things as animated 2D line charts over time, so they will dynamically adjust to new maximums.  No more pies.

**Reply by iamumairayub** [reply]
Thanks again for reply.

> 10 GB / 30 GB is a third.

Oh my bad.

Any ETA on when Cronicle v2 will be released?

**Reply by jhuckaby** [reply]
It should be early next year (2024), but I can't really promise anything.  Life keeps getting in the way 😊
